tests/test_spark.py

```diff
@@ -0,0 +1,146 @@
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+
+import logging
+from time import sleep
+from typing import Dict, Tuple
+
+import pytest
+import requests
+from docker import DockerClient
+from docker.models.containers import Container
+from granulate_utils.metrics import MetricsSnapshot
+from granulate_utils.metrics.metrics import (
+    SPARK_AGGREGATED_STAGE_METRICS,
+    SPARK_APPLICATION_DIFF_METRICS,
+    SPARK_APPLICATION_GAUGE_METRICS,
+    SPARK_EXECUTORS_METRICS,
+)
+from granulate_utils.metrics.mode import SPARK_STANDALONE_MODE
+from granulate_utils.metrics.sampler import BigDataSampler
+from granulate_utils.metrics.spark import SparkRunningApps
+from pytest import LogCaptureFixture
+
+from gprofiler.log import get_logger_adapter
+from tests.conftest import build_image
+from tests.utils import wait_container_to_start
+
+# `BigDataSampler` receives a logger as an argument.
+logger = get_logger_adapter("gprofiler_spark_test")
+
+# List that includes all the metrics that are expected to be collected by `BigDataSampler` in SA mode.
+EXPECTED_SA_METRICS_KEYS = [
+    metric
+    for metrics_dict in (
+        SPARK_APPLICATION_GAUGE_METRICS,
+        SPARK_APPLICATION_DIFF_METRICS,
+        SPARK_AGGREGATED_STAGE_METRICS,
+        SPARK_EXECUTORS_METRICS,
+    )
+    for metric in metrics_dict.values()
+]
+
+SPARK_MASTER_HOST = "127.0.0.1"
+
+DISCOVER_INTERVAL_SECS = 5
+DISCOVER_TIMEOUT_SECS = 60
+
+
+@pytest.fixture
+def sparkpi_container(docker_client: DockerClient) -> Container:
+    """
+    This fixture is responsible for running SparkPi application in a container.
+    See `containers/spark/Dockerfile`
+    """
+    spark_image = build_image(docker_client=docker_client, runtime="spark")
+    container = docker_client.containers.run(
+        spark_image,
+        detach=True,
+        mounts=[],
+        network_mode="host",
+        pid_mode="host",
+        environment={"SPARK_MASTER_HOST": SPARK_MASTER_HOST},
+    )
+    wait_container_to_start(container)
+    try:
+        yield container
+    finally:
+        container.stop()
+        container.remove()
+
+
+def _wait_for_sparkpi_to_start() -> None:
+    """
+    Waits for SparkPi to be recognized by Master.
+    Doing so using `SparkRunningApps` class, `get_running_apps()` method.
+
+    Because Master is provisioning, it takes a few seconds until we receive a response from it.
+    """
+    running_apps = SparkRunningApps(SPARK_STANDALONE_MODE, f"http://{SPARK_MASTER_HOST}:8080", logger)
+    apps: Dict[str, Tuple[str, str]] = {}
+    while not apps:
+        try:
+            apps = running_apps.get_running_apps()
+        except requests.exceptions.ConnectionError as e:
+            if "Max retries exceeded" in str(e):
+                # Spark Master is not ready yet.
+                pass
+            sleep(1)
+
+
+def _validate_spark_sa_metricssnapshot(snapshot: MetricsSnapshot) -> None:
```

Comment:
Can you test the values as well? For example, number of apps, number of executors, ...

Reply:
6ae02d5